File: src/main/java/ru/hollowhorizon/hc/common/world/storage/HollowWorldData.java
package ru.hollowhorizon.hc.common.world.storage;

import net.minecraft.nbt.CompoundNBT;
import net.minecraft.world.storage.WorldSavedData;
import ru.hollowhorizon.hc.client.utils.NBTUtils;
import ru.hollowhorizon.hc.common.world.structures.StoryStructureData;

import java.util.ArrayList;

public class HollowWorldData extends WorldSavedData {
    public static HollowWorldData INSTANCE;
    public ArrayList<StoryStructureData> STRUCTURE_DATA_LIST = new ArrayList<>();

    public HollowWorldData() {
        super("hollow_world_data");
    }

    @Override
    public void load(CompoundNBT nbt) {
        STRUCTURE_DATA_LIST = NBTUtils.loadList(nbt, "story_structures", StoryStructureData.SERIALIZER);
    }

    @Override
    public CompoundNBT save(CompoundNBT nbt) {
        NBTUtils.saveList(nbt, "story_structures", STRUCTURE_DATA_LIST, StoryStructureData.SERIALIZER);
        return nbt;
    }

    @Override
    public boolean isDirty() {
        return true;
    }
}

File: src/main/java/ru/hollowhorizon/hc/common/world/storage/HollowWorldData.kt
package ru.hollowhorizon.hc.common.world.storage

import net.minecraft.nbt.CompoundNBT
import net.minecraft.world.storage.WorldSavedData
import ru.hollowhorizon.hc.client.utils.nbt.NBTFormat
import ru.hollowhorizon.hc.client.utils.nbt.deserialize
import ru.hollowhorizon.hc.client.utils.nbt.serialize
import ru.hollowhorizon.hc.common.world.structures.StoryStructureData

class HollowWorldData : WorldSavedData("hollow_world_data") {
    @JvmField
    var STRUCTURE_DATA_LIST = ArrayList<StoryStructureData>()
    override fun load(nbt: CompoundNBT) {
        val structureData = nbt.get("structures") ?: return

        STRUCTURE_DATA_LIST = NBTFormat.deserialize(structureData)
    }
    override fun save(nbt: CompoundNBT): CompoundNBT {
        nbt.put("structures", NBTFormat.serialize(STRUCTURE_DATA_LIST))

        return nbt
    }

    override fun isDirty(): Boolean {
        return true
    }

    companion object {
        @JvmField
        var INSTANCE: HollowWorldData? = null
    }
}
